OptimoveNotificationServiceExtension/Sources/CategoryManager.swift

```diff
@@ -1,92 +1,55 @@
 //  Copyright © 2022 Optimove. All rights reserved.
 
-import Foundation
 import UserNotifications
 
-let MAX_DYNAMIC_CATEGORIES = 128
-let DYNAMIC_CATEGORY_IDENTIFIER = "__kumulos_category_%d__"
-
-@available(iOS 10.0, *)
-class CategoryManager {
-    let categoryReadLock = DispatchSemaphore(value: 0)
-    let dynamicCategoryLock = DispatchSemaphore(value: 1)
-
-    fileprivate static var instance: CategoryManager?
-
-    static var sharedInstance: CategoryManager {
-        if instance == nil {
-            instance = CategoryManager()
-        }
-
-        return instance!
+enum CategoryManager {
+    enum Constants {
+        static let MAX_DYNAMIC_CATEGORIES = 128
+        static let DYNAMIC_CATEGORY = OptimobileUserDefaultsKey.DYNAMIC_CATEGORY.rawValue
     }
 
-    static func getCategoryIdForMessageId(messageId: Int) -> String {
-        return String(format: DYNAMIC_CATEGORY_IDENTIFIER, messageId)
+    static func getCategoryId(messageId: Int) -> String {
+        return "__kumulos_category_\(messageId)__"
     }
 
-    static func registerCategory(category: UNNotificationCategory) {
-        var categorySet = sharedInstance.getExistingCategories()
-        var storedDynamicCategories = sharedInstance.getExistingDynamicCategoriesList()
+    static func registerCategory(_ category: UNNotificationCategory) async {
+        var systemCategories = await UNUserNotificationCenter.current().notificationCategories()
+        var storedCategoryIds = readCategoryIds()
+
+        systemCategories.insert(category)
+        storedCategoryIds.insert(category.identifier)
 
-        categorySet.insert(category)
-        storedDynamicCategories.append(category.identifier)
+        let (categories, categoryIds) = maybePruneCategories(categories: systemCategories, categoryIds: storedCategoryIds)
 
-        sharedInstance.pruneCategoriesAndSave(categories: categorySet, dynamicCategories: storedDynamicCategories)
+        UNUserNotificationCenter.current().setNotificationCategories(categories)
+        writeCategoryIds(categoryIds)
 
         // Force a reload of the categories
-        _ = sharedInstance.getExistingCategories()
+        await UNUserNotificationCenter.current().notificationCategories()
     }
 
-    private func getExistingCategories() -> Set<UNNotificationCategory> {
-        var returnedCategories = Set<UNNotificationCategory>()
-
-        UNUserNotificationCenter.current().getNotificationCategories { (categories: Set<UNNotificationCategory>) in
-            returnedCategories = Set<UNNotificationCategory>(categories)
-
-            self.categoryReadLock.signal()
-        }
-
-        _ = categoryReadLock.wait(timeout: DispatchTime.now() + DispatchTimeInterval.seconds(5))
```

Comment:
Not sure about this deletion. 
This extension is a separate process that has a limited amount of time (30 seconds) to finish the current execution. If the time exceeds, [another](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension/1648227-serviceextensiontimewillexpire) method will be called for this instance, and we will have the last chance to call the callback the [first call method](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension/1648229-didreceive).